### PR TITLE
ref(*): move proxy-service mapping from catalog to proxy registry

### LINKED ISSUE #3272: Move GetServicesForProxy out of MeshCatalog and into Proxy Registry / Proxy{}
This GitHub issue is to move `GetServicesForProxy()` out of `MeshCataloger` interface and into the `Proxy{}` struct.

This is an improved version of the https://github.com/openservicemesh/osm/issues/3149 issue.

---

With #3229 we changed from
```go
GetServicesFromEnvoyCertificate(certificate.CommonName) ([]service.MeshService, error)
```
to
```go
GetServicesForProxy(*envoy.Proxy) ([]service.MeshService, error)
```
which is streamlined and easy to understand.

We have an opportunity to further carve out the `Proxy Registry` / `Endpoint Discovery` / `Service Discovery` out of the Mesh Catalog. (I believe the proxy, endpoint, service discovery and registries are connected, sometimes overlapping).

---
In a graph representation we would have one `Proxy` connected to 0, 1, or more `MeshService` objects.
---

With this it would make sense to further evolve
```go
GetServicesForProxy(*envoy.Proxy) ([]service.MeshService, error)
```

to be

```go
Proxy{}.GetServices() ([]service.MeshService, error)
```
---

### How do we get there?
- [ ] We could and should move `GetServicesForProxy` out of `MeshCatalog` and closer to the `ProxyRegistry` (prototype #3114): https://github.com/openservicemesh/osm/blob/0e9429bb9c44bb6b026719d895449e534c66a1ad/pkg/catalog/xds_certificates.go#L19-L20
It seems that in this move we could make `GetServicesForProxy()` be a member of `Prox{}`. Then `Proxy` would need to have a `kubeController`. We could pass a link to kubeController at the time of `proxy` initialization: https://github.com/openservicemesh/osm/blob/0e9429bb9c44bb6b026719d895449e534c66a1ad/pkg/envoy/ads/stream.go#L45
(We would need to let the ADS Server struct be aware of the kube controller too. That's ok - that ADS responds to Envoys on that particular Kubernetes cluster. But longer term this may be too specific. At some point Envoys could be connecting from outside of Envoy. These would have a different cert, indicating the origin of where they are coming from.)
- [ ] We can also **rename** (at that point) this function to be `Proxy{}.GetServices()`
- [ ] Use this in the places where `GetServicesForProxy()` was used
  - [ ] CDS: https://github.com/openservicemesh/osm/blob/0e9429bb9c44bb6b026719d895449e534c66a1ad/pkg/envoy/cds/response.go#L17
  - [ ] RDS:  https://github.com/openservicemesh/osm/blob/0e9429bb9c44bb6b026719d895449e534c66a1ad/pkg/envoy/rds/response.go#L29
  - [ ] LDS: https://github.com/openservicemesh/osm/blob/0e9429bb9c44bb6b026719d895449e534c66a1ad/pkg/envoy/lds/response.go#L21

---

Here is a very rough draft of what this may look like: #3275

### PR DESCRIPTION
**Pull request by nojnhuh**
**Description**:
This change factors out the logic mapping an Envoy proxy instance to the
services it backs from the `MeshCatalog` into the `ProxyRegistry`.

Most of the individual changes are a result of moving declarations
around. The most substantial changes is a new interface defined in
`pkg/envoy/registry`:

    // ProxyServiceMapper knows how to map Envoy instances to services.
    type ProxyServiceMapper interface {
            ListProxyServices(*envoy.Proxy) ([]service.MeshService, error)
    }

This change includes two implementations: `KubeProxyServiceMapper` for
Kubernetes and `ExplicitProxyServiceMapper` for custom cases (only used
in unit tests currently).

No functional changes.

Fixes #3272

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

**Comment by nojnhuh** [comment]
I split this up as much as I could, so hopefully this is easier to review. The first three commits are pretty much just moving things around to accommodate the last commit, which has the most significant changes. I skipped signing off the first commits just to remind myself to squash before merging.

**Review by jaellio — APPROVED** [review]

File: pkg/envoy/registry/services.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+// ProxyServiceMapper knows how to map Envoy instances to services.
+type ProxyServiceMapper interface {
+	ListProxyServices(*envoy.Proxy) ([]service.MeshService, error)
+}
+
+// ExplicitProxyServiceMapper is a custom ProxyServiceMapper implementation.
+type ExplicitProxyServiceMapper func(*envoy.Proxy) ([]service.MeshService, error)
```
Comment (shashankram) [inline pkg/envoy/registry/services.go:21]: Why is this needed?

File: pkg/envoy/registry/services.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+// ListProxyServices executes the given mapping.
+func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
+	return e(p)
+}
+
+// KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
+type KubeProxyServiceMapper struct {
+	KubeController k8s.Controller
+}
+
```
Comment (shashankram) [inline pkg/envoy/registry/services.go:32]: I am not in favor of exposing k8s constructs under `pkg/envoy`. We have tried to get away from this as much as possible (ex. service account -> service identity changes).

Is this avoidable?

File: pkg/envoy/registry/services.go (hunk 1 above)
Comment (nojnhuh) [reply pkg/envoy/registry/services.go:21]: Some of the unit tests in `pkg/envoy/*ds` use a mocked `MeshCatalog` that doesn't expose any of the Kubernetes details required for `KubeProxyServiceMapper` to work.

e.g.
https://github.com/openservicemesh/osm/blob/25744ba1a1259f5863b627862850cc1333abf1c0/pkg/envoy/cds/response_test.go#L46-L48

File: pkg/envoy/registry/services.go (hunk 2 above)
Comment (nojnhuh) [reply pkg/envoy/registry/services.go:32]: That makes sense. I can't say for certain it's unavoidable, I'll see if I can refactor this.

File: pkg/envoy/registry/services.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
+func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
```
Comment (shashankram) [inline pkg/envoy/registry/services.go:34]: I would imagine `GetServices()` to be implemented o the `envoy.Proxy` object. The `KubeProxyServiceMapper` seems avoidable given it only encapsulates `kubernetes.Controller`.

File: pkg/envoy/registry/services.go (hunk 1 above)
Comment (shashankram) [reply pkg/envoy/registry/services.go:21]: I left a related comment below, but I wonder why this intermediary KubeProxyServiceMapper struct is even necessary. Can't we have the call flow as `XDS -> proxy.GetServices() -> kubeController.Something()`?

File: cmd/osm-controller/osm-controller.go [hunk 4]
```diff
@@ -215,7 +215,7 @@ func main() {
 		cfg,
 		endpointsProviders...)
 
-	proxyRegistry := registry.NewProxyRegistry()
+	proxyRegistry := registry.NewProxyRegistry(&registry.KubeProxyServiceMapper{KubeController: kubernetesClient})
```
Comment (draychev) [inline cmd/osm-controller/osm-controller.go:218]: This seems like a nice abstraction.  You have the ProxyRegistry, and then you have this other facility, which knows how to get the services for a given proxy.

File: pkg/envoy/registry/services.go (hunk 3 above)
Comment (draychev) [reply pkg/envoy/registry/services.go:34]: That's not a bad idea. 
So instead of `service, err := proxyRegistry.ListProxyServices(proxy)` we can have `service, err := proxy.GetServices()`?

@nojnhuh if you also agree this is a good simplification - perhaps feel free to open a follow-up GitHub issue to re-refactor this in another PR.

File: pkg/envoy/registry/services.go (hunk 2 above)
Comment (draychev) [reply pkg/envoy/registry/services.go:32]: @shashankram What would you recommend? I am curious what you have in mind?

I agree on keeping k8s out of `pkg/envoy`, but also it'd be tough to determine services for envoy (pod) without leveraging k8s.  So this boils down to perhaps a different package (location) where the `ProxyServiceMapper` implementation lives?

I would not be opposed to @nojnhuh doing this in another PR if he prefers to do that.

**Review by draychev — APPROVED** [review]
@nojnhuh, this looks great!  Thank you for working on carving out these from the MeshCatalog.

I like @shashankram's proposals. It seems to me this PR here is already good to go, and we can open follow-up issues and iterate - to avoid frustrations around merge conflicts etc.

File: pkg/envoy/registry/services.go (hunk 2 above)
Comment (shashankram) [reply pkg/envoy/registry/services.go:32]: @nojnhuh and I chatted a bit about this. None of my comments need to be resolved in this specific PR, though @nojnhuh wanted to play around with refactoring the code a bit more to explore the feasibility of my comments.

File: pkg/envoy/registry/services.go (hunk 1 above)
Comment (nojnhuh) [reply pkg/envoy/registry/services.go:21]: If `proxy.GetServices()` always invoked a `kubeController`, then those tests I mentioned above will break since the tests mock everything instead of creating kube pods and services (even with the fake client). Those tests could probably be rewritten to introduce a bunch of kubernetes concepts, but that would make them much harder to grok.

Since we need two ways to map proxies to services (kube and non-kube), we have an interface with two types implementing that behavior, and the proxy has a handle to an instance of one of those types. That's why we have that extra level of indirection.

File: pkg/envoy/registry/services.go (hunk 2 above)
Comment (nojnhuh) [reply pkg/envoy/registry/services.go:32]: > I am not in favor of exposing k8s constructs under pkg/envoy.

Having played with this a bit, it seems like if this is moved into pkg/kubernetes, then we're exposing Envoy details there which doesn't seem any better. I suppose we could create another package for this, but that doesn't really solve the problem of mixing kuberentes and envoy details.

To really fix this it seems like we'd need a set of implementation-agnostic proxy abstractions so this could then map kubernetes services to abstract proxies instead of Envoy proxies.

File: pkg/envoy/registry/services.go (hunk 3 above)
Comment (nojnhuh) [reply pkg/envoy/registry/services.go:34]: > I would imagine `GetServices()` to be implemented o the `envoy.Proxy` object.

That should be simple to change. My main reasoning for doing it this way was that potential related functionality (perhaps `ListServiceProxies()`) makes more sense to me implemented in a layer above a single `Proxy` and maintaining that related behavior in different places seems less than ideal. In this case I can see where `proxy.GetServices()` is more ergonomic though.

> The `KubeProxyServiceMapper` seems avoidable given it only encapsulates `kubernetes.Controller`.

Do you propose this functionality be implemented directly against `kubernetes.Controller`? That would involve exposing Envoy details there, which I mentioned in the other thread.

Comment (shashankram) [reply pkg/envoy/registry/services.go:34]: Don't want to complicate this, let's merge this :-)

**Review by shashankram — APPROVED** [review]
@nojnhuh would you mind squashing the commits and rebasing to HEAD.

**Comment by nojnhuh** [comment]
Thank you for the reviews and your patience @jaellio, @draychev, and @shashankram!
